### ISSUE TITLE
[!!!][FEATURE] New TypoScript parser

### ISSUE THREAD

**Issue by TYPO3IncTeam** [TYPO3-Documentation/Changelog-To-Doc#116]
:information_source: View this commit [on Github](https://github.com/TYPO3/typo3/commit/09c61884bf2f15b0f209f8b177268620fc77cc8a)
:busts_in_silhouette: Authored by Christian Kuhn lolli@schwarzbu.ch
:heavy_check_mark: Merged by Christian Kuhn lolli@schwarzbu.ch

## Commit message

[!!!][FEATURE] New TypoScript parser

This adds a rewritten TypoScript syntax parser to ext:core.

The heart of the existing parser dates back to Kasper's
initial commit in 2003 and most parts have never been
touched structurally. Even though various parts of
TypoScript became less important in the Frontend rendering
chain over time, it still plays a central point. Looking at
the given parser approach, it became clear the project can
benefit from a revamped parser based on modern PHP code.

Goals:
* A well structured, flexible and tested codebase.
* Fix long standing syntax shenanigans and limits of
  the current parser approach and make the syntax
  interpretation more robust and resilient.
* Have a better cache layer that kicks in more often
  to ultimately gain speed building the FE TypoScript.
* Allow improving the Backend "Template" module to
  show and analyze more TypoScript details.
* Have an object based structure for the final TypoScript
  tree, and allow an array export of it to keep backwards
  compatibility.

The new implementation aims to fully substitute the old
parser. This first patch however only brings the "library"
part of needed code changes: There is no usage of the new
parser whatsoever with this patch, except of course unit
and functional tests.

To understand the structure of the new parser, it might be
helpful to have a rough understanding of the old parser. We'll
look at the frontend parsing steps only, Backend usages in
Template module are similar, though.

When accessing a frontend page, TYPO3 at some points finds the
requested page uid. It then builds the "rootline" of page
records from the top page root node down to the requested
page. This creates a list of page database records, mountpoints
and similar are taken care of at this point, too.

This rootline is fed to the old TypoScript parser. The parser now
finds all relevant sys_template records attached to the rootline
pages. These are the "entry" TypoScript snippets to parse. It
then resolves all "includes": Snippets from sys_template
include_static_file db records, snippets from various globals,
snippets from sys_template basedOn, and so on. It also resolves
@import and <INCLUDE_TYPOSCRIPT: syntax and substitutes them
with the included content. All that is gathered in one huge
string. The main parsing process then goes through this string
and creates the TypoScript array, while taking care of conditions
and constants substitution at the same time.

This approach has various drawbacks: First, gathering everything
into one main string forces to parse the entire string a-new for
each and every page: A different page could have different
includes or sys_template records attached. Secondly, parsing
conditions and constants while creating the final array from the
source string, ties "runtime" information (especially conditions)
hard into the main parsing process, rendering an effective cache
layer impossible. Third, the main strategy of gathering everything
in one string leads to various funny details, for instance when
opening a condition in one file and not closing it with [end] or
[global], this state leaks to the next file. Bracket handling "}"
has similar issues.

These structural issues can't be changed by refactoring the
given codebase, the only option is to rewrite the entire thing.

IncludeTree:
The entry point to the new parser is IncludeTree/TreeBuilder:
This one again receives the pages "rootline" from Frontend. But
instead of creating one huge string, it creates an object tree of
includes. Each attached sys_template record is a child of
"RootInclude" in the tree, and each sys_template include node
can have children for further includes (like include_static_file).
Single source snippets of each include are tokenized (see below)
and analyzed for @import and <INCLUDE_TYPOSCRIPT: and conditions:
If they exist, an include is marked as "split". Each part is then
represented by child includes. We thus receive a tree of objects
where imports are already resolved, and conditions are child nodes
in the tree.

The main advantage is this tree does not carry runtime information.
A single snippet *always* leads to the same tree, no matter if a
condition matched or not. This way, the tree can be cached and the
full tree, or parts of it can be re-used when requesting a
different page. To do that efficiently, a new cache "typoscript"
is established to store a serialized representation of the tree
as php file, which can be unserialized relatively quickly.

Tokenizer:
The tokenizer takes a single source snipped and creates a stream
of Line objects from it, with tokens representing the source line.
For instance, a TypoScript snipped like "foo = fooValue" creates
a IdentifierAssignmentLine object, with a T_IDENTIFIER token for
"foo" and a T_VALUE token for "fooValue". There are Line classes
for all the different Lines (assignment, copy, condition, ...) and
Tokens for the various details (T_VALUE, T_CONSTANT, ...). Tokens
are sometimes encapsulated in TokenStreams, for example multiple
identifiers "foo.bar = barValue" create a stream of the two
T_IDENTIFIER tokens "foo" and "bar". This depends on the line type.

We also have two different tokenizers: The Frontend uses the
LossyTokenizer which creates a stream only of "relevant" tokens:
Empty lines, comments and lines with invalid syntax are ignored.
The LosslessTokenizer however creates a 1:1 representation of the
source snipped, including token positions (line and start column).
This allows detailed analysis in the Backend Template module.

AstBuilder:
This is the third part of the structure: A representation of
parsed TypoScript as object tree. This is similar to the array
we're dealing with now, but it can carry additional functionality.
For instance, the well-known TypoScript array can be extracted
from it.
Using the AstBuilder works like this: Create the IncludeTree
first (from cache). Next, traverse the IncludeTree with a visitor
that looks at condition includes and sets an information
if they match. Then, feed all includes that matched to the
AstBuilder to create the object tree. The AstBuilder receives
a LineStream of each include and extends the object tree depending
on the line type.

This means: IncludeTree building and source tokenizing can be
cached in Frontend, applying constants and condition verdicts
together with building the finally TypoScript object tree is a
runtime step on each page. Separating tokenizing and AST building
with a cache layer in between is roughly twice as fast in
Frontend compared to the previous solution.

The patch establishes these three main structures and comes with
an extensive set of unit and functional tests. It is an extract
of a bigger WIP patch that already contains usages of the new
structure in Frontend and some parts of the Backend. The patch
also adds a ReST file to explain some subtle TypoScript syntax
changes. These will kick in as soon as we start using the
structure with upcoming changes.

The new parser should be rather robust already. Further changes
that start using the new parser will probably only change minor
things. For now, the entire structure is marked @internal,
though: API is still the array representation, only. This will
change later when we actively start using the TypoScript object
tree in the core.

Change-Id: I4047a878494078b6ac149553fa305c1d69329e37
Resolves: #97816
Resolves: #96503
Resolves: #90146
Resolves: #41327
Resolves: #76447
Releases: main
Reviewed-on: https://review.typo3.org/c/Packages/TYPO3.CMS/+/74987
Tested-by: core-ci <typo3@b13.com>
Tested-by: Benni Mack <benni@typo3.org>
Tested-by: Stefan Bürk <stefan@buerk.tech>
Tested-by: Christian Kuhn <lolli@schwarzbu.ch>
Reviewed-by: Benni Mack <benni@typo3.org>
Reviewed-by: Stefan Bürk <stefan@buerk.tech>
Reviewed-by: Christian Kuhn <lolli@schwarzbu.ch>

## :heavy_plus_sign: Added files

<details>
<summary>12.0/Breaking-97816-TypoScriptSyntaxChanges.rst</summary>

```rst
.. include:: /Includes.rst.txt

.. _breaking-97816-1656350406:

============================================
Breaking: #97816 - TypoScript syntax changes
============================================

See :issue:`97816`

Description
===========

.. note::

    The new TypoScript parser is currently not or only partially active
    in the system. The full implementation should be finished until v12 LTS.
    This note should be removed later.

TYPO3 v12 comes with a new TypoScript syntax parser that is more performant,
more robust and allows better tooling in the Backend.

The new parser is more forgiving in many places, but some seldom used syntax
details have been removed, too. This documentation explains details that may
be breaking for existing instances.

Also see :doc:`the feature documentation <Feature-97816-TypoScriptSyntaxImprovements.rst>`
for an overview of syntax improvements.

Impact
======

Using one of the constructs below stops working in v12 and needs
TypoScript adaptions.


Affected installations
======================

Instances using TypoScript as outlined below.


Migration
=========

Streamlined constants usage
---------------------------

It has never been fully documented in which context "constants" :typoscript:`{$foo}`
shall be used and which exact capabilities they have. The main TypoScript constants
documentation within the :ref:`TypoScript Reference <t3tsref:Constants>` was partially
outdated, and the :ref:`TSconfig documentation <t3tsconfig:Syntax>` claimed TSconfig
is not constants aware at all, which isn't fully the case anymore. Let's sort out
some details:

* Nesting constants is **not** possible and never has been. A construct like
  this is invalid syntax and is treated as string literal: :typoscript:`{$foo{$bar}}`

* Recursive constants were possible with the old parser but are not supported with the new
  parser anymore. This was never documented, the Backend Template module never showed them as
  resolved, only the Frontend parsed recursive constants. The simple rule is now: Never
  access a constant within another constant. Instances using a construct like the below one
  need to untie constants.

  .. code-block:: typoscript

    constants:
    foo = fooValue
    # This does not resolve to "fooValue" but is kept as string literal "{$foo}"
    bar = {$foo}

    setup:
    # This does NOT resolve to "fooValue", but to the string literal "{$foo}"
    myValue = {$bar}

* Constants are now restricted to "assignments" and "conditions". Using a constant to
  substitute an "identifier" / "object path" is no longer allowed. This has never been
  clarified in the docs before and instances abusing constants to specify object paths
  should be seldom and need to resolve the situation with the new parser now:

  This is supported:

  .. code-block:: typoscript

    # Simple constant usage as assignment value:
    foo = {$bar}
    # Compiling a value with string literals and constants:
    foo = I am {$bar}
    # Using a constant in a condition:
    [ myValue = {$bar} ]
    # Using constant(s) in multiline assginments:
    foo (
       I am {$bar} and {$baz}
    )

  These constructs are *not* supported:

  .. code-block:: typoscript

    # Using a constant as object path specification
    {$bar} = myValue
    # This is an object path specification, too, and not supported:
    foo < {$bar}

* TSconfig *does* support constant substitution: Site constants can be used
  in TSconfig. This has been introduced with TYPO3 v10, see
  :doc:`<../10.4/Feature-91080-SiteSettingsAsTsConstantsAndInTsConfig.rst>` for details.


UTF-8 BOM in TypoScript files
-----------------------------

The new TypoScript parser no longer ignores `UTF-8 BOM <https://en.wikipedia.org/wiki/Byte_order_mark>`_
in included files: Having a Byte-order-mark in TypoScript files may create undesired
results. They should be removed. UTF-8 BOM is disallowed in various other languages,
for instance JSON and PHP. The new parser follows here. Modern editors typically don't
add an UTF-8 BOM anymore.

Instances can check if they use UTF-8 BOM with a Unix shell command:

.. code-block:: bash

    # find affected files
    find . -type f -print0 | xargs -0 -n1 file {} | grep 'UTF-8 Unicode (with BOM)'
    # remove UTF-8 BOM from a single file
    sed -i '1s/^\xEF\xBB\xBF//' affectedFile.typoscript


Support for \n and \r\n linebreaks only
---------------------------------------

TypoScript sources must terminate single lines with either "\n" (Unix ending: LineFeed),
or "\r\n" (Windows ending: Carriage return, LineFeed). Ancient Mac, prior to Mac OS X
used "\r" as single linebreak character. This old linebreak type is no longer detected
when parsing TypoScript and may lead to funny results, but chances are very low any
instance is affected by this.


.. index:: Backend, Frontend, TSConfig, TypoScript, NotScanned, ext:core

```

</details>

<details>
<summary>12.0/Feature-97816-TypoScriptSyntaxImprovements.rst</summary>

```rst
.. include:: /Includes.rst.txt

.. _feature-97816-1656350667:

================================================
Feature: #97816 - TypoScript syntax improvements
================================================

See :issue:`97816`

Description
===========

.. note::

    The new TypoScript parser is currently not or only partially active
    in the system. The full implementation should be finished until v12 LTS.
    This note should be removed later.

TYPO3 v12 comes with a new TypoScript syntax parser that is more performant,
more robust and allows better tooling in the Backend.

The new parser is more forgiving in many places, this documentation lists
the new capabilities.

Also see :doc:`the breaking documentation <Breaking-97816-TypoScriptSyntaxImprovements.rst>`
for an overview of breaking syntax changes.


Impact
======

Improved comment parsing
------------------------

TypoScript comment detection had various nasty quirks with the old parser. The confusing
behavior did lead to many headaches in the past and not sticking to the weird parser
restrictions in comment parsing could easily lead to unexpected results, often ignoring
bigger sections of the subsequent TypoScript lines.

This has been relaxed heavily: Comment detection should almost always act as developers
and integrators would expect from a language. Especially the former obligation to place
a closing multiline comment (:typoscript:`*/`) on a single line to close the comment
section has been removed.

A couple of examples to clarify:

.. code-block:: typoscript

  foo # This is a comment to an invalid line

  foo < bar // This is a comment
  foo < bar /* This is a valid comment, too */

  foo > # Another valid comment

  foo := addToList(1) # Yes, a comment

  [foo = bar] # Much comment. Much wow.

  <INCLUDE_TYPOSCRIPT: source="..."> /* A comment */

  foo (
    # This is NOT a comment but part of the value assignment!
    bar = barValue
  ) # This is a comment

  foo = bar // This is NOT a comment but part of the value assignment!

@import in conditions
---------------------

Placing a :typoscript:`@import` keyword within a condition is now supported,
the example below works. Note this obsoletes the clumsy :typoscript:`<INCLUDE_TYPOSCRIPT:`
syntax, and integrators are encouraged to fully switch to :typoscript:`@import`.

.. code-block:: typoscript

  [loginUser('*')]
    @import 'EXT:my_extension/Configuration/TypoScript/LoggedInUser.typoscript
  [ELSE]
    @import 'EXT:my_extension/Configuration/TypoScript/NotLoggedInUser.typoscript
  [END]

Scope restriction to file / snipped level
-----------------------------------------

The old TypoScript parser merged the entire TypoScript for a page into one big
chunk of text. The new parser does not do that anymore, but parses each included
snippet one-by-one. This automatically means state no longer leaks to subsequent
snippets. Missing closing brackets :typoscript:`}` in one file do not destroy block
integrity of a following include anymore. Same for conditions: A missing closing
condition block (:typoscript:`[END]` or :typoscript:`[GLOBALS]`) no longer leaks
to another file - a conditions ends at the end of a file or snippet.

Nesting conditions is partially supported
-----------------------------------------

Nesting conditions is partially possible with the new TypoScript parser, **if** the
conditions are in different files. As example, let's first sort what happens when
two conditions follow directly in one snippet:

.. code-block:: typoscript

  [loginUser('*')]
    @import 'EXT:my_extension/Configuration/TypoScript/LoggedInUser.typoscript
  [applicationContext == "Development"]
    @import 'EXT:my_extension/Configuration/TypoScript/Development.typoscript
  [END]

This always worked and did not change with the new parser: Opening a new condition
automatically closes the preceding one. In the example above, both conditions are
standalone: :file:`Development.typoscript` is included no matter if a user is
logged in or not.

But, and this in new, nesting conditions within different files is possible now.
In the example below, file :file:`LoggedInUserDevelopment.typoscript` is only
included if a user is logged in *and* the application is in development context.

.. code-block:: typoscript

  [loginUser('*')]
    @import 'EXT:my_extension/Configuration/TypoScript/LoggedInUser.typoscript
  [END]

  File LoggedInUser.typoscript:
  [applicationContext == "Development"]
    @import 'EXT:my_extension/Configuration/TypoScript/LoggedInUserDevelopment.typoscript
  [END]

Irrelevant order of <INCLUDE_TYPOSCRIPT: tokens
----------------------------------------------

The :typoscript:`<INCLUDE_TYPOSCRIPT:` keywords has the three properties
:typoscript:`source`, :typoscript:`condition` and :typoscript:`extensions`. They had to
be in a specific order with the old parser, but can be placed in arbitrary order now.

Further clarifications
----------------------

* The "reference" :typoscript:`=<` operator is not a direct language construct. The parser
  understands the syntax, but does not resolve it. Allowed :typoscript:`=<` operator are very
  limited: In general, it can *only* be used for Frontend Content Objects, typically like this:
  :typoscirpt:`tt_content.bullets =< lib.contentElement`. Another usage is referencing :typoscript:`lib.parseFunc`.
  Using the :typoscript:`=<` operator in these cases can have a performance advantage since it
  avoids an expensive copy operation that is done lazily if really needed. There are two methods
  that resolve this operator, namely :php:`ContentObjectRenderer->cObjGetSingle()` and
  :php:`ContentObjectRenderer->mergeTSRef()`.
  This also means: The :typoscript:`=<` operator is *not supported* in TypoScript constants,
  is only supported for specific elements in TypoScript setup, and is *not supported* in TSconfig.
  Also note the reference operator does not support "relative" copies like the "copy" operator supports
  with :typoscript:`20 < .10` and similar.

* The new parser has a minor change in behavior with the "copy" :typoscript:`<` operator on top-level.
  This shouldn't have huge impact in real life usage and is documented for completeness. Consider this
  example:

  .. code-block::

    lib.viewConfig {
      baz = bazValue
    }

    first = FLUIDTEMPLATE
    first < lib.viewConfig

  The situation is there that :typoscript:`lib.viewConfig` has no assigned value (just children). The
  target :typoscript:`first` however has value :typoscript:`FLUIDTEMPLATE`. The old TypoScript parser
  usually keeps the "target" value in such cases, but only if the TypoScript object is not on top level
  (:typoscript:`first` in contrast to :typoscript:`first.10` or similar). In the example above, value
  :typoscript:`FLUIDTEMPLATE` would vanish with the old parser, but is now kept with the new parser.

.. index:: Backend, Frontend, TSConfig, TypoScript, ext:core

```

</details>

**Comment by linawolf** [comment]
Put on hold until the new parser is better integrated into the Core. 

Get ok from @lolli42  before this gets worked on or merged.

**Comment by brotkrueml** [comment]
Mind also the removed introductory note: https://github.com/TYPO3-Documentation/Changelog-To-Doc/issues/351
